app/models/activity_tag.rb

```diff
@@ -14,4 +14,49 @@ class ActivityTag < ApplicationRecord
   belongs_to :activity
 
   validates :activity_id, uniqueness: { scope: %i[tag_type_id tag_id] }
+  validate :validate_tag_matches_type
+  validate :validate_tag_parent
+  validate :validate_tag_child
+
+  private
+
+  def validate_tag_matches_type
+    return if tag.nil?
+
+    return if tag.tag_type == tag_type
+
+    errors.add :tag_id, I18n.t('errors.tag.should_match_tag_type')
+  end
+
+  def validate_tag_parent
+    return if correct_parent_tag?
+
+    errors.add :tag_id, I18n.t('errors.tag.should_match_parent')
+  end
+
+  def validate_tag_child
+    return if correct_child_tag?
+
+    errors.add :tag_id, I18n.t('errors.tag.should_match_parent')
+  end
+
+  def correct_parent_tag?
+    return true if tag.nil? || tag.tag_type.parent.nil?
+
+    # e.g., if Category is DCC, is this a DCC Subcategory?
+    parent_tag == tag.parent
+  end
+
+  def correct_child_tag?
+    # e.g., if this is a DCC Category, is the child a DCC Subcategory?
+    child_tag.nil? || tag.children.include?(child_tag)
```

Comment:
Can get undefined method `children' for nil:NilClass here if add a new activity then set the parent to None